### PR TITLE
Rename `Microsoft.Terminal.TerminalControl` to `.Control`; Split into dll & lib

### PR DESCRIPTION
**Pull request by zadjii-msft**
**BE NOT AFRAID**. I know that there's 107 files in this PR, but almost
all of it is just find/replacing `TerminalControl` with `Control`.

This is the start of the work to move TermControl into multiple pieces,
for #5000. The PR starts this work by:
* Splits `TerminalControl` into separate lib and dll projects. We'll
  want control tests in the future, and for that, we'll need a lib.
* Moves `ICoreSettings` back into the `Microsoft.Terminal.Core`
  namespace. We'll have other types in there soon too. 
  * I could not tell you why this works suddenly. New VS versions? New
    cppwinrt version? Maybe we're just better at dealing with mdmerge
    bugs these days.
* RENAMES  `Microsoft.Terminal.TerminalControl` to
  `Microsoft.Terminal.Control`. This touches pretty much every file in
  the sln. Sorry about that (not sorry). 

An upcoming PR will move much of the logic in TermControl into a new
`ControlCore` class that we'll add in `Microsoft.Terminal.Core`.
`ControlCore` will then be unittest-able in the
`UnitTests_TerminalCore`, which will help prevent regressions like #9455 

## Detailed Description of the Pull Request / Additional comments
You're really gonna want to clean the sln first, then merge this into
your branch, then rebuild. It's very likely that old winmds will get
left behind. If you see something like 

```
Error    MDM2007    Cannot create type
Microsoft.Terminal.TerminalControl.KeyModifiers in read-only metadata
file Microsoft.Terminal.TerminalControl.
```

then that's what happened to you.

### CODE REVIEW

**Comment by DHowett** [comment]
If we're gonna rename the namespace, now is as good a time as any to rename the DLL file too. Microsoft.Terminal.Control.dll.

**Comment by DHowett** [comment]
(WinRT consumers may contractually (apparently) try to resolve namespace X.Y.Z by doing LoadLibrary(X.Y.Z.dll), X.Y.dll, and X.dll)

**Comment by DHowett** [comment]
I still don't understand the thesis. If we make the interactivity stuff part of Core, _that's what we test_. Are we doing work today (make the UWP Control a lib) that we'll be _undoing_ later?

**Comment by zadjii-msft** [comment]
> If we're gonna rename the namespace, now is as good a time as any to rename the DLL file too. Microsoft.Terminal.Control.dll.

yea okay



> Are we doing work today (make the UWP Control a lib) that we'll be _undoing_ later?

probably not? We might be adding tests in the future - when we do, we'll be happy that we have a lib. But we _will_ be moving most of the testable code into M.T.Core.lib, which is already easily testable, so that's not the biggest worry. What I am worried about is that the delicate house of cards we have will fall over if I revert part of it. I did the dll->dll&lib split first, then did the Control.ICoreSettings -> Core.ICoreSettings thing. Part of me is worried that will no longer work if I revert the lib&dll split

File: src/cascadia/TerminalApp/TerminalAppLib.vcxproj [hunk 1]
```diff
@@ -346,6 +346,12 @@
       <Private>false</Private>
       <CopyLocalSatelliteAssemblies>false</CopyLocalSatelliteAssemblies>
     </Reference>
+    <Reference Include="Microsoft.Terminal.Core">
+      <HintPath>$(OpenConsoleCommonOutDir)TerminalCore\Microsoft.Terminal.Core.winmd</HintPath>
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalAppLib.vcxproj:350]: augh another hardcoded winmd reference

File: src/cascadia/TerminalApp/dll/TerminalApp.vcxproj [hunk 2]
```diff
@@ -75,6 +75,28 @@
...
+    MDMERGE will know where the TermControl types are defined. However, we need
+    to do it exactly like this so the packaging project won't roll up
+    TermControl's .xbf's from both the TermControl project and this one. -->
+<!--     <Reference Include="Microsoft.Terminal.Control">
+      <HintPath>$(OpenConsoleCommonOutDir)TerminalControl\Microsoft.Terminal.Control.winmd</HintPath>
+      <IsWinMDFile>true</IsWinMDFile>
+      <Private>false</Private>
+      <CopyLocalSatelliteAssemblies>false</CopyLocalSatelliteAssemblies>
+    </Reference> -->
+
```
Comment (DHowett) [inline src/cascadia/TerminalApp/dll/TerminalApp.vcxproj:90]: This comment spells doom and gloom if we _don't_ do this thing, but then it's commented out. Why do we NOT need it?

File: src/cascadia/TerminalControl/TerminalControlLib.vcxproj [hunk 3]
```diff
@@ -2,27 +2,31 @@
...
-    <ProjectName>TerminalControl</ProjectName>
-    <RootNamespace>Microsoft.Terminal.TerminalControl</RootNamespace>
-    <!-- cppwinrt.build.pre.props depends on these settings: -->
-    <!-- build a dll, not exe (Application) -->
-    <ConfigurationType>DynamicLibrary</ConfigurationType>
+    <Keyword>Win32Proj</Keyword>
+    <ProjectName>TerminalControlLib</ProjectName>
+    <RootNamespace>Microsoft.Terminal.Control</RootNamespace>
+    <TargetName>Microsoft.Terminal.ControlLib</TargetName>
+    <WindowsTargetPlatformMinVersion>10.0.17763.0</WindowsTargetPlatformMinVersion>
```
Comment (DHowett) [inline src/cascadia/TerminalControl/TerminalControlLib.vcxproj:9]: **Hold up**. We have this in a common location.

File: src/cascadia/TerminalControl/TerminalControlLib.vcxproj [hunk 4]
```diff
@@ -110,61 +126,22 @@
...
+  <!-- ====================== Compiler & Linker Flags ===================== -->
   <ItemDefinitionGroup>
-    <Link>
-      <AdditionalDependencies>dwrite.lib;dxgi.lib;d2d1.lib;d3d11.lib;shcore.lib;winmm.lib;pathcch.lib;propsys.lib;uiautomationcore.lib;Shlwapi.lib;ntdll.lib;user32.lib;kernel32.lib;%(AdditionalDependencies)</AdditionalDependencies>
-    </Link>
     <ClCompile>
+      <PrecompiledHeaderFile>pch.h</PrecompiledHeaderFile>
       <AdditionalIncludeDirectories>$(OpenConsoleDir)src\cascadia\inc;$(OpenConsoleDir)src\types\inc;%(AdditionalIncludeDirectories)</AdditionalIncludeDirectories>
     </ClCompile>
+    <Link>
```
Comment (DHowett) [inline src/cascadia/TerminalControl/TerminalControlLib.vcxproj:135]: libs do not run link.exe.

Comment (DHowett) [reply src/cascadia/TerminalControl/TerminalControlLib.vcxproj:135]: (remove this.)

File: src/cascadia/TerminalControl/TerminalControlLib.vcxproj [hunk 5]
```diff
@@ -110,61 +126,22 @@
...
+      <AdditionalDependencies>dwrite.lib;dxgi.lib;d2d1.lib;d3d11.lib;shcore.lib;winmm.lib;pathcch.lib;propsys.lib;uiautomationcore.lib;Shlwapi.lib;ntdll.lib;user32.lib;kernel32.lib;%(AdditionalDependencies)</AdditionalDependencies>
+      <!-- <AdditionalDependencies>WindowsApp.lib;user32.lib;shell32.lib;%(AdditionalDependencies)</AdditionalDependencies> -->
```
Comment (DHowett) [inline src/cascadia/TerminalControl/TerminalControlLib.vcxproj:137]: and this

File: src/cascadia/TerminalControl/TerminalControlLib.vcxproj [hunk 6]
```diff
@@ -110,61 +126,22 @@
...
-      <ReferencePath Include="@(OCReferencePathToRestore)"/>
-    </ItemGroup>
-  </Target>
-  <Target Name="OpenConsoleStripDuplicateWinmdFromReferencesBeforePass2">
-    <ItemGroup>
-      <!-- Nobody else uses this item group, it's safe to never restore. -->
-      <WinMDReferenceToCompile Remove="@(WinMDReferenceToCompile)" Condition="'%(Filename)' == '$(RootNamespace)'"/>
-    </ItemGroup>
-  </Target>
+  <Import Project="$(SolutionDir)build\rules\CollectWildcardResources.targets" />
```
Comment (DHowett) [inline src/cascadia/TerminalControl/TerminalControlLib.vcxproj:146]: shocked that resources might just "work", scared too

File: src/cascadia/TerminalControl/dll/TerminalControl.vcxproj [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+    <!-- C++/WinRT sets the depth to 1 if there is a XAML file in the project
+         Unfortunately for us, we need it to be 3. When the namespace merging
+         depth is 1, Microsoft.Terminal.Control becomes "Microsoft",
+         and our WinMD file becomes "Microsoft". Because WinRT is very
+         namespace-driven, this winmd is considered to contain the entire
+         Microsoft namespace. This is, obviously, not great. None of our other
+         projects compile properly when they depend on this "Microsoft.winmd."
+    -->
+    <CppWinRTNamespaceMergeDepth>3</CppWinRTNamespaceMergeDepth>
```
Comment (DHowett) [inline src/cascadia/TerminalControl/dll/TerminalControl.vcxproj:23]: we should not be doing any namespace merging in this DLL file -- are we?

File: src/cascadia/TerminalControl/dll/pch.cpp [hunk 8]
```diff
@@ -0,0 +1,4 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "pch.h"
```
Comment (DHowett) [inline src/cascadia/TerminalControl/dll/pch.cpp:4]: do we need one of these? can we just .. not have them at all?

File: src/cascadia/WindowsTerminal/WindowsTerminal.vcxproj [hunk 9]
```diff
@@ -83,6 +83,19 @@
     <ProjectReference Include="$(OpenConsoleDir)src\types\lib\types.vcxproj" />
   </ItemGroup>
 
+  <!-- Manually add a reference to Core here. We need this so MDMERGE will know
+  where the Core types are defined. However, we need to do it exactly like this,
+  because the Core project is a lib not a dll, so it can't be ProjectReference'd
+  like everything else. It does get built into TerminalControl.dll, so we don't
+  need to worry about that. -->
+  <ItemGroup>
+    <Reference Include="Microsoft.Terminal.Core">
```
Comment (DHowett) [inline src/cascadia/WindowsTerminal/WindowsTerminal.vcxproj:92]: probably need one in WindowsTerminalUniversal too

**Review by DHowett — CHANGES_REQUESTED** [review]

File: src/cascadia/TerminalApp/dll/TerminalApp.vcxproj (hunk 2 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/dll/TerminalApp.vcxproj:90]: I'm gonna be real with you - no idea why this works now. maybe it thinks the .xbfs belong to TerminalControlLib now, not TerminalControl.dll?

File: src/cascadia/TerminalControl/dll/pch.cpp (hunk 8 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalControl/dll/pch.cpp:4]: all the other libs have them just like this, totally empty. IIRC the project gets mad if it has _absolutely nothing_ to compile, so you gotta give it something (with nothing in it) to make it happy.

File: src/cascadia/TerminalControl/TerminalControlLib.vcxproj (hunk 6 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalControl/TerminalControlLib.vcxproj:146]: me too

File: OpenConsole.sln [hunk 10]
```diff
@@ -1518,6 +1529,33 @@ Global
...
+		{CA5CAD1A-5555-4AC7-AC72-6CA5B3AB89ED}.Debug|x64.Build.0 = Debug|x64
+		{CA5CAD1A-5555-4AC7-AC72-6CA5B3AB89ED}.Debug|x86.ActiveCfg = Debug|Win32
+		{CA5CAD1A-5555-4AC7-AC72-6CA5B3AB89ED}.Debug|x86.Build.0 = Debug|Win32
+		{CA5CAD1A-5555-4AC7-AC72-6CA5B3AB89ED}.Release|Any CPU.ActiveCfg = Release|Win32
+		{CA5CAD1A-5555-4AC7-AC72-6CA5B3AB89ED}.Release|ARM.ActiveCfg = Release|Win32
+		{CA5CAD1A-5555-4AC7-AC72-6CA5B3AB89ED}.Release|ARM64.ActiveCfg = Release|ARM64
+		{CA5CAD1A-5555-4AC7-AC72-6CA5B3AB89ED}.Release|ARM64.Build.0 = Release|ARM64
+		{CA5CAD1A-5555-4AC7-AC72-6CA5B3AB89ED}.Release|DotNet_x64Test.ActiveCfg = Release|Win32
+		{CA5CAD1A-5555-4AC7-AC72-6CA5B3AB89ED}.Release|DotNet_x86Test.ActiveCfg = Release|Win32
+		{CA5CAD1A-5555-4AC7-AC72-6CA5B3AB89ED}.Release|x64.ActiveCfg = Release|x64
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply OpenConsole.sln:1555]: I straight up copied TerminalControl's original lines here. If this is wrong, how will I know?

[the remark this answers is not in the document]
Comment (zadjii-msft) [reply OpenConsole.sln:1555]: it looks like `*|DotNet_x64Test.*`, `*|DotNet_x86Test.*`, and `*|Any CPU.*` all always map to `Win32`. The Audit lines are all weird:

```
		{CA5CAD1A-F542-4635-A069-7CAEFB930070}.AuditMode|Any CPU.ActiveCfg = Debug|Win32
		{CA5CAD1A-F542-4635-A069-7CAEFB930070}.AuditMode|ARM.ActiveCfg = AuditMode|Win32
		{CA5CAD1A-F542-4635-A069-7CAEFB930070}.AuditMode|ARM64.ActiveCfg = Release|ARM64
		{CA5CAD1A-F542-4635-A069-7CAEFB930070}.AuditMode|DotNet_x64Test.ActiveCfg = Debug|Win32
		{CA5CAD1A-F542-4635-A069-7CAEFB930070}.AuditMode|DotNet_x86Test.ActiveCfg = Debug|Win32
		{CA5CAD1A-F542-4635-A069-7CAEFB930070}.AuditMode|x64.ActiveCfg = Release|x64
		{CA5CAD1A-F542-4635-A069-7CAEFB930070}.AuditMode|x86.ActiveCfg = Release|Win32
```

presumably though should all map to either Audit, Debug or Release, but presumably all the same thing.

Then again, TerminalAppLib is the same way:

```
		{CA5CAD1A-9A12-429C-B551-8562EC954746}.AuditMode|Any CPU.ActiveCfg = Debug|Win32
		{CA5CAD1A-9A12-429C-B551-8562EC954746}.AuditMode|ARM.ActiveCfg = AuditMode|Win32
		{CA5CAD1A-9A12-429C-B551-8562EC954746}.AuditMode|ARM64.ActiveCfg = Release|ARM64
		{CA5CAD1A-9A12-429C-B551-8562EC954746}.AuditMode|DotNet_x64Test.ActiveCfg = Debug|Win32
		{CA5CAD1A-9A12-429C-B551-8562EC954746}.AuditMode|DotNet_x86Test.ActiveCfg = Debug|Win32
		{CA5CAD1A-9A12-429C-B551-8562EC954746}.AuditMode|x64.ActiveCfg = Release|x64
		{CA5CAD1A-9A12-429C-B551-8562EC954746}.AuditMode|x86.ActiveCfg = Release|Win32
``` 

so maybe this is fine

**Review by DHowett — APPROVED** [review]
108/108, looks fine to me. Mechanical. Makes our build times slower, I suspect, but what doesn't these days? Thanks for doing my renames. :smile:

**Comment by DHowett** [comment]
@microsoft/windows-console-team this is a mechanical change -- can we get it reviewed?

**Comment by DHowett** [comment]
Body and title edited to fit commit style (ignore me)

**Review by carlos-zamora — APPROVED** [review]

**Pull request merged**